### PR TITLE
Script for saving audio augmented by Lhotse samplers/dataloaders

### PR DESCRIPTION
**Pull request by racoiaws**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a script that saves audio produced by a dataloader with a Lhotse sampler inside, mimicking training setup. This allows for easier debugging of augmentation pipelines, as well as preparing offline validation sets that would use same augmentations as ones used at training time.

**Collection**: Audio

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

The script piggybacks on an existing model config.
```python
python examples/audio/save_augmented.py \
    +input_cuts=some_path/cuts.jsonl \
    +output_cuts=some_other_path/cuts.gsm_and_clipping_augmented.jsonl \
    +keep_directory_structure=true \
    model.sample_rate=48000 \
    ++model.train_ds.rir_enabled=true \
    ++model.train_ds.rir_path=path/to/rir_manifest.jsonl
```
We can add or override augmentation options via CLI in the usual - whatever is currently supported by sampler.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Needs to be updated when https://github.com/NVIDIA-NeMo/NeMo/pull/14322 gets merged

### CODE REVIEW

File: examples/audio/save_augmented.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+Assumptions:
+- input data are described as a Lhotse CutSet in a JSONL file
+   - consists of simple MonoCuts with Recording paths relative to the Cuts manifest
+- the parent directory of the output cuts must exist
+
+Requires additional config parameters `input_cuts` and `output_cuts`.
+Produces:
+- %output_cuts_parent_dir%/audio/
+- %output_cuts_parent_dir%/%output_cuts_filename%.jsonl
+where the audio folder contains the augmented and clean signals, respectively, with `.dirty.flac` and `.clean.flac` suffixes.
```
Comment (anteju) [inline examples/audio/save_augmented.py]:
```
`.dirty.flac` and `.clean.flac``
```
If these correspond to input and output cuts, then name the files accordingly as `.input.flac` and `.output.flac`.

Alternatively, `.input.flac` and `.augmented.flac`.

File: examples/audio/save_augmented.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+If `keep_directory_structure` provided and is True, the script will preserve the directory structure of the input cuts.
+
+Text is preserved from the input cuts if possible. 
+
+Optional config parameter `num_samples` can be used to limit the number of samples to save (but not more than input dataloader size). 
+If not specified, the dataloader is used until exhausted.
+"""
+
+
+def check_input_cuts(input_cuts_path: Path) -> None:
```
Comment (anteju) [inline examples/audio/save_augmented.py:64]: Add docstring.

File: examples/audio/save_augmented.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    ), "output_cuts is required, please override (for example, +output_cuts=some_path/cuts.augmented.jsonl)"
+    num_samples = cfg.get("num_samples", None)
+    sample_rate = cfg.model.sample_rate
+    keep_directory_structure = cfg.get("keep_directory_structure", False)
+
+    input_cuts_path = Path(cfg.input_cuts)
+    output_cuts_path = Path(cfg.output_cuts)
+    check_input_cuts(input_cuts_path)  # throws an exception if they aren't ok
+
+    assert output_cuts_path.parent.exists(), "output_cuts parent directory must exist"
```
Comment (anteju) [inline examples/audio/save_augmented.py]: Print `output_cuts_path` or parent.

File: examples/audio/save_augmented.py (hunk 1 above)
Comment (anteju) [reply examples/audio/save_augmented.py]: Also, update naming in `main` accordingly.

File: examples/audio/save_augmented.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    OmegaConf.update(cfg, "model.train_ds.cuts_path", str(input_cuts_path), force_add=True)
+    OmegaConf.update(cfg, "model.train_ds.shuffle", False)  # ensure deterministic behavior
+    OmegaConf.update(cfg, "model.train_ds.batch_size", 1)
+    OmegaConf.update(cfg, "model.train_ds.shard_seed", 0, force_add=True)  # ensure deterministic behavior
+    if cfg.model.train_ds.get("sample_rate", None) != sample_rate:
+        OmegaConf.update(cfg, "model.train_ds.sample_rate", sample_rate, force_add=True)
+
+    dataloader = get_lhotse_dataloader_from_config(
+        OmegaConf.create(cfg.model.train_ds), global_rank=0, world_size=1, dataset=LhotseAudioToTargetDataset()
+    )
```
[the remark this answers is not in the document]
Comment (racoiaws) [reply examples/audio/save_augmented.py:131]: No, SimpleCutSampler can't be used as a drop-in replacement while keeping the full logic from `get_lhotse_sampler_from_config`.

There are three API mismatches:

No constraint parameter — SimpleCutSampler hardcodes a TimeConstraint(max_duration, max_cuts) internally. But `get_lhotse_sampler_from_config` calls `determine_sampling_constraint()` at line 625, which returns a `SamplingConstraint` object (supporting multimodal sampling, 2D bucketing, etc.) and passes it to the sampler. `SimpleCutSampler` can't accept that.

No shuffle_buffer_size — SimpleCutSampler uses a traditional full-shuffle-at-epoch-start (via DataSource), not the streaming reservoir-sampling shuffle of DynamicCutSampler.

No quadratic_duration — used for transformer-like models with quadratic complexity.

The pre-sampler `CutSet` transforms (channel selection, resampling, augmentations, filtering) and the post-sampler `.map()` transforms (lowpass, clipping, RIR, compression) would all work fine — those are sampler-agnostic. It's only the sampler constructor itself that's incompatible.
DynamicCutSampler with shuffle=False already preserves input order. Looking at its `__iter__`:

```
        seed = resolve_seed(self.seed)
        self.cuts_iter = [iter(cs) for cs in self.cuts]
        if self.shuffle:
            self.cuts_iter = [
                streaming_shuffle(
                    cs,
                    rng=random.Random(seed + self.epoch),
                    bufsize=self.shuffle_buffer_size,
                )
                for cs in self.cuts_iter
            ]
        # ... feeds into DurationBatcher which reads sequentially
```


There **is** a potential issue. When `shuffle=False`, it's just `iter(cs)` — purely sequential. The reordering risk comes only from `DynamicBucketingSampler` (which sorts into duration buckets).

The fix for `save_augmented.py` is twofold:

1. Force `use_bucketing=False` to prevent reordering
2. Disable all filters so no cuts are silently dropped (preserving the positional zip)

[the remark this answers is not in the document]
Comment (racoiaws) [reply examples/audio/save_augmented.py:131]: The second part might be a bit excessive but why not.

File: examples/audio/save_augmented.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    assert input_cuts_path.exists(), "input_cuts must exist"
+    assert input_cuts_path.suffix == '.jsonl', "input_cuts must be a .jsonl file"
+    assert input_cuts_path.parent.exists(), "input_cuts parent directory must exist"
+    cuts = lhotse.CutSet.from_file(input_cuts_path)
+    for i, cut in enumerate(cuts):
+        assert isinstance(cut, MonoCut), f"{i}th cut is a {type(cut)}, not a MonoCut"
+        assert len(cut.recording.sources) == 1, f"{i}th cut has {len(cut.recording.sources)} sources"
+        assert cut.recording.sources[0].source is not None, "{i}th cut has no audio source specified"
+
+        recording_path = Path(cut.recording.sources[0].source)
```
[the remark this answers is not in the document]
Comment (racoiaws) [reply examples/audio/save_augmented.py:75]: Correct observation, incorrect suggested change (missed by 2 lines)

File: examples/audio/save_augmented.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        "max_duration",
+        "min_tps",
+        "max_tps",
+        "min_tokens",
+        "max_tokens",
+        "max_cer",
+        "min_context_speaker_similarity",
+    ):
+        OmegaConf.update(cfg, f"model.train_ds.{key}", getattr(_defaults, key), force_add=True)
+
```
[the remark this answers is not in the document]
Comment (racoiaws) [reply examples/audio/save_augmented.py:128]: `LhotseAudioToTargetDataset` does not return ids and we are not changing that for this script

**Review by anteju — APPROVED** [review]
